### PR TITLE
Make offline compute compatible with scream

### PR DESCRIPTION
**Pull request by elynnwu**
In this PR, we added the capability of taking native scream model output and running it through `fv3fit.train` and `fv3net.diagnostics.offline.compute`.

Added public API:
- `format_output.py`: this is a post processing script for native scream output. Currently, we also include making random tendencies values so that it can go through training and offline diagnostics workflow.
- `open_nudge_to_fine_scream`: similar to `open_nudge_to_fine` but specifically for scream
- `gsrm_name_from_resolution_string`: new function in vcm to determine whether to use fv3 or scream based on resolution string

Refactored public API:
- `BatchesFromMapperConfig` now has `ptop`. Defaults are set to fv3 setting, but can be changed in the input namelist
- `offline/compute.py`: `_get_transect` takes in `ptop` as input
- DiagArg: now has `horizontal_dims` and `vertical_dim`

Significant internal changes:
- `vcm/calc/thermo/constants.py`: changed `TOA_PRESSURE` to public variable so that it can be used in other package
- catalog: add scream ne30 grid and land sea mask
- Changed `HORIZONTAL_DIMS` to `HORIZONTAL_DIMS_FV3`

- [x] Tests added

Coverage reports (updated automatically):

### CODE REVIEW

File: external/loaders/loaders/batches/_batch.py
Comment (oliverwm1) [inline external/loaders/loaders/batches/_batch.py:96]: Why provide `gsrm_name` as a config option if we just overwrite it based on `res`? If we can just use the existing `res` parameter to describe the horizontal grid, that would be better.

File: external/loaders/loaders/mappers/_nudged/_nudged.py
Comment (oliverwm1) [inline external/loaders/loaders/mappers/_nudged/_nudged.py:188]: How much control do we have over the SCREAM output names? Do you know what the names of these quantities (physics tendencies) will be? 

And if they're not named `tendency_of_air_temperature_due_to_scream_physics`, probably we should just not make this change. Note that this function does not fail if these variables don't exist in the given dataset.

File: external/vcm/vcm/catalog.yaml
Comment (oliverwm1) [inline external/vcm/vcm/catalog.yaml:1012]: I wonder if we should save this `area` data in units of `m^2` instead of the current `rad^2`? High potential for bugs, although I understand a desire to be consistent with SCREAM output.

Comment (oliverwm1) [reply external/vcm/vcm/catalog.yaml:1012]: Also another question: did you reshape this data to 6 x 90 x 40? Or is that the native output shape?

File: external/vcm/vcm/catalog.yaml
Comment (oliverwm1) [inline external/vcm/vcm/catalog.yaml:1021]: I don't *think* it will cause issues, but just FYI the FV3 land sea mask includes a sea-ice mask (value=2), whereas it seems the SCREAM one is just 0 (ocean) or land (1).

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:155]: We should probably define this convention somewhere centralized (in `vcm`?) instead of having similar functions that distinguish between scream and fv3 "res" strings in many places.

I'd also suggest checking `res.startswith("ne")` and `res.startswith("c")` instead of `in`. (Maybe another grid will come along with a `c` in it somewhere!)

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:175]: This is a bit scary... do we expect user to remember to do this every time they load the "area" data for SCREAM? In practice this catalog item gets used a lot in notebooks and other places.

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:177]: the added `ptop` argument is not used in this function.

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:31]: Is this used anywhere?

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:32]: I guess this has some standard meaning for SCREAM, but I have no idea what it represents. Either give a more useful variable name (preferred) or add a short comment.

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:163]: Why stack and then unstack?

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:99]: The standard name for omega is `lagrangian_tendency_of_air_pressure`. `upward_air_velocity` might get confused with the vertical velocity in units of m/s.

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:85]: I don't follow rationale for this definition of `x_dim` and `y_dim`. Let's chat offline.

File: workflows/post_process_run/scream_post/format_output.py
Comment (oliverwm1) [inline workflows/post_process_run/scream_post/format_output.py:1]: How would this script get called? Are you imagining some automated post-processing, or more of an ad-hoc approach? Will the script require a lot of memory to run for our anticipated datasets?

For now, this script may be better suited for the `fv3net/projects` location, which is a good place for untested code. Everything else in `post_process_run` is quite well unit tested and more modular so I'm not sure this script is currently a good fit.

File: external/loaders/loaders/batches/_batch.py
Comment (oliverwm1) [inline external/loaders/loaders/batches/_batch.py:147]: Same point as above in terms of `res` and `gsrm_name` arguments.

**Review by oliverwm1 — CHANGES_REQUESTED** [review]
Great that we can get model training/offline diagnostics compatible with scream!

Could we get away without adding a new `gsrm_name` config parameter?

**Comment by elynnwu** [comment]
Sample native scream output: `gs://vcm-ml-scratch/elynnw/scream/output.scream.monthly.AVERAGE.nmonths_x1.0001-02-01-00000.nc`

File: workflows/post_process_run/scream_post/format_output.py
Comment (spencerkclark) [inline workflows/post_process_run/scream_post/format_output.py:161]: Is it necessary to assign these coordinates, or can we get away without it?

File: workflows/post_process_run/scream_post/format_output.py
Comment (spencerkclark) [inline workflows/post_process_run/scream_post/format_output.py:57]: Is this method more accurate than using `hyai`, `hybi`, and `ps` to compute the pressure on the layer interfaces, because SCREAM is a non-hydrostatic model?  It's too bad they don't have this output as a diagnostic directly...

More generally it would probably be good to have this written in more xarray-style code, e.g. without the assumption of array shapes and position of the vertical dimension.

File: workflows/post_process_run/scream_post/format_output.py
Comment (spencerkclark) [inline workflows/post_process_run/scream_post/format_output.py:23]: Is this not included in the dataset?

File: workflows/post_process_run/scream_post/format_output.py
Comment (spencerkclark) [reply workflows/post_process_run/scream_post/format_output.py:85]: Also interested in this just for my own understanding.

**Review by spencerkclark — COMMENTED** [review]
Thanks @elynnwu -- I think @oliverwm1 covered the important things.  I had a few more minor comments as I looked this over.

File: external/loaders/loaders/batches/_batch.py
Comment (elynnwu) [reply external/loaders/loaders/batches/_batch.py:96]: We provide `gsrm_name` inside `mapper_config`. This is because functions like `open_nudge_to_fine` does not know anything about resolution but does need the name to read in the correct physics tendency variables. 
Then in the general config, instead of adding another `gsrm_name`, we use `res` to determine `gsrm_name`.

File: external/loaders/loaders/mappers/_nudged/_nudged.py
Comment (elynnwu) [reply external/loaders/loaders/mappers/_nudged/_nudged.py:188]: Apparently, it is possible for us to change scream output names (in the future). Currently there isn't any physics tendency variables, so I'm keeping the same naming convention here when making fake data.

File: external/vcm/vcm/catalog.yaml
Comment (elynnwu) [reply external/vcm/vcm/catalog.yaml:1021]: Right, we only have 0 and 1 in this land sea mask. I'm waiting to swap this to scream's land sea mask when I have access to it, not sure if that one will have sea ice.

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (elynnwu) [reply workflows/diagnostics/fv3net/diagnostics/offline/compute.py:177]: It actually needs to be passed to `interpolate_to_pressure_levels`, updated now

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:31]: Nope, deleted

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:32]: No longer need these

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:163]: No reshaping

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:99]: No renaming for now

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:85]: These were some magic integers I came up with, but per our discussion, no more reshaping. We use native scream output shape.

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:1]: Moved to `fv3net/projects/scream`. This script is called once offline to transform scream native output to zarr, it also includes some hacks to work with the current set of scream output. We expect to remove these hacks (e.g., surface pressure is 2x, horizontal winds are written in the same variable but will be split into 2) once we have the new scream output. Wanted to still commit this so that the dataset is reproducible

File: external/loaders/loaders/batches/_batch.py
Comment (elynnwu) [reply external/loaders/loaders/batches/_batch.py:147]: Responded above

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:161]: We can get way without it, removed.

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:57]: In the future, we should have `delp` directly as part of the output. Changed to using `hyai`, `hybi`, `p0`, and `ps` to calculate the full pressure.

File: workflows/post_process_run/scream_post/format_output.py
Comment (elynnwu) [reply workflows/post_process_run/scream_post/format_output.py:23]: Not currently, but it should be soon.

**Review by spencerkclark — COMMENTED** [review]
Thanks for the updates @elynnwu!  Just a question / suggestion related to time.  I'll leave it to @oliverwm1 to review the new dimension name code and give the final approval since that was something you both discussed.

File: external/loaders/loaders/batches/_batch.py
Comment (oliverwm1) [inline external/loaders/loaders/batches/_batch.py:213]:
```suggestion
        transforms.append(add_grid_info(res))
        if vcm.model_name_from_resolution_string(res) == "fv3":
            transforms.append(add_wind_rotation_info(res)])
```

Comment (oliverwm1) [reply external/loaders/loaders/batches/_batch.py:213]: Could do something like this (and define appropriate function in `vcm`). It seems like we want to do this `res` -> model name transformation in various places, so we should probably make a single function to do it, and then use that function elsewhere.

File: workflows/diagnostics/fv3net/diagnostics/offline/_input_sensitivity.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/_input_sensitivity.py:25]: Is there a way to pass the horizontal dims to this function? So could just modify it to be `stack_dims = horizontal_dims + [DATASET_DIM_NAME]` etc. instead of needing a new if statement.

File: workflows/diagnostics/fv3net/diagnostics/_shared/constants.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/_shared/constants.py:30]: Is the plan to use `z` or `lev` for the vertical dimension? There are still some places where `z` is hard-coded. Probably better to leave it hard-coded everywhere (fine with me) or make sure it is actually configurable.

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:155]: Should consolidate this logic to one place (e.g. a `vcm` function).

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:316]:
```suggestion
```

File: workflows/diagnostics/fv3net/diagnostics/offline/compute.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute.py:320]:
```suggestion
        evaluation_resolution=evaluation_grid.sizes[horizontal_dims[0]],
```
And infer `horizontal_dims` somewhere earlier in code. And then can also pass `horizontal_dims` to `_compute_diagnostics` instead of inferring it within that function.

File: workflows/diagnostics/fv3net/diagnostics/offline/compute_diagnostics.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/compute_diagnostics.py:68]: I suggest inferring the `horizontal_dims` once in the main function of the compute code, and then passing it down to this function. Will be less liable to bugs and easier to make changes if the dim name convention changes for SCREAM.

**Review by oliverwm1 — COMMENTED** [review]
Thanks for the changes @elynnwu! We're getting closer. Happy to discuss on call if easier.

My main suggestions are:
1) write a new mapper function for opening SCREAM data for ML (sorry, should have realized this approach was best on the last review)
2) consolidate logic for determining if a "res" string corresponds to FV3 or SCREAM to one place
3) determine horizontal dimensions once, and then pass around to the functions that need them, instead of repeating code to infer the dimension names.

File: external/loaders/loaders/mappers/_nudged/_nudged.py
Comment (elynnwu) [reply external/loaders/loaders/mappers/_nudged/_nudged.py:129]: This makes sense, and it would alleviate a lot of the gsrm name issue.

File: workflows/diagnostics/fv3net/diagnostics/_shared/constants.py
Comment (elynnwu) [reply workflows/diagnostics/fv3net/diagnostics/_shared/constants.py:30]: The plan is to use `z`, retained `VERTICAL_DIM="z"` and reused it here.

File: workflows/diagnostics/fv3net/diagnostics/offline/_input_sensitivity.py
Comment (elynnwu) [reply workflows/diagnostics/fv3net/diagnostics/offline/_input_sensitivity.py:25]: Done. Needs to be sorted because `horizontal_dims` is `["x","y","tile"]` but here it needs to be tile first.

Comment (oliverwm1) [reply workflows/diagnostics/fv3net/diagnostics/offline/_input_sensitivity.py]: Probably would be fine to just define `horizontal_dims` with the tile-first order.

File: external/loaders/loaders/mappers/_nudged/_nudged.py
Comment (oliverwm1) [inline external/loaders/loaders/mappers/_nudged/_nudged.py:283]: Do you know where in the time loop nudging happens for SCREAM? The purpose of this code is to make sure that the ML inputs are the atmospheric state right before nudging. It's possible this code will need to change if the nudging is implemented differently than what we have in FV3GFS.

File: external/vcm/vcm/catalog.yaml
Comment (oliverwm1) [inline external/vcm/vcm/catalog.yaml:1023]: this zarr currently has `lndgrid` as the dimension name. Should it be `ncol`?

File: workflows/diagnostics/fv3net/diagnostics/prognostic_run/metrics.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/prognostic_run/metrics.py:157]: I think we should really try to avoid these deep `if` statements. Since this PR is just to make the offline diagnostics work for SCREAM, I suggest just leaving this hard-coded to `HORIZONTAL_DIMS_FV3`. Same for other cases in this module.

File: external/vcm/vcm/convenience.py
Comment (oliverwm1) [inline external/vcm/vcm/convenience.py:164]: I would delete this function, and instead just have code in `fv3net/diagnostics` to return the appropriate dim names based on the output from `gsrm_name_from_resolution_string`. Right now, we've defined the FV3 horizontal dims in two places (here, and as `HORIZONTAL_DIMS_FV3` in `fv3net/diagnostics/_shared/constants.py`, which is not ideal.

File: workflows/diagnostics/fv3net/diagnostics/_shared/constants.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/_shared/constants.py:6]: additionally define the SHIELD horizontal dims here.

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (oliverwm1) [inline workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:177]: should we just rename in the saved zarr?

**Review by oliverwm1 — COMMENTED** [review]
Looking good, just a few minor comments.

File: external/loaders/loaders/mappers/_nudged/_nudged.py
Comment (elynnwu) [reply external/loaders/loaders/mappers/_nudged/_nudged.py:283]: Right, it is actually flexible in scream where nudging happens (the order of any process is specified in the namelist). I'm keeping it like this because we do not actually have a nudged run yet, but I added a todo note to update this once we have it solidified.

File: external/vcm/vcm/catalog.yaml
Comment (elynnwu) [reply external/vcm/vcm/catalog.yaml:1023]: It should but I do not have the atmosphere's land mask and this is to help me remember that I need to update this once I get that file. TODO note in `load_grid_info_scream`

File: workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py
Comment (elynnwu) [reply workflows/diagnostics/fv3net/diagnostics/offline/_helpers.py:177]: Yes, I don't want to mix it here because the file needs to be updated to the atmosphere's mask instead of using land model's. Will open a separate PR as soon as I get that file.

**Review by oliverwm1 — APPROVED** [review]
LGTM! Make sure to update the PR description so it is accurate with current status of PR, and use that as the commit message. There is also a typo in the PR title (compatiable -> compatible).

If you'd like (optional!) you can enable the integration test before merging. It's possible this will catch something new.

**Comment by elynnwu** [comment]
Integration test worked on the 3rd attempt w/o OOM error, timed out due to the 35 min limit.

**Pull request merged**
